### PR TITLE
Refactor ConcurrentBoundedQueue

### PR DESCRIPTION
**Pull request by kitaisreal**
Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Closes #29720.

### CODE REVIEW

File: src/Common/ZooKeeper/ZooKeeperImpl.cpp [hunk 1]
```diff
@@ -840,9 +840,9 @@ void ZooKeeper::finalize(bool error_send, bool error_receive)
 
     auto expire_session_if_not_expired = [&]
     {
-        /// No new requests will appear in queue after close()
-        bool was_already_closed = requests_queue.close();
-        if (!was_already_closed)
+        /// No new requests will appear in queue after finish()
+        bool was_already_finished = requests_queue.finish();
```
Comment (tavplubix) [inline src/Common/ZooKeeper/ZooKeeperImpl.cpp:844]: We must not clear `requests_queue` here, see https://github.com/ClickHouse/ClickHouse/blob/fe6b7c77c7d6bd2a45a20f3b6bb4eb91da6177ff/src/Common/ZooKeeper/ZooKeeperImpl.cpp#L952-L955

**Review by alesapin — COMMENTED** [review]
Nice refactoring, thx.

**Review by alesapin — APPROVED** [review]
Approve but need to fix @tavplubix comment

**Review by tavplubix — APPROVED** [review]

**Comment by tavplubix** [comment]
Integration tests failures look suspicious

**Review by alesapin — APPROVED** [review]
Now should be Ok.

**Comment by kitaisreal** [comment]
Test failures are not related to changes.
Functional tests: 01288_shard_max_network_bandwidth - flacky.
Integration tests: test_MemoryTracking - flacky.

**Pull request merged**